### PR TITLE
Wt 21 implement business logic

### PR DESCRIPTION
**Pull request by k1mka**

### CODE REVIEW

File: lib/data/repository/repository.dart [hunk 1]
```diff
@@ -0,0 +1,3 @@
+abstract class Repository {
+  Future<void> registerUser(String password, String email);
```
Comment (zddarova) [inline lib/data/repository/repository.dart:2]: are you sure about Future<void> reutrn type? mby we need to return Future<bool> for case if user registered or not? or something else, void is too uninformative

File: lib/data/services/registration_service.dart [hunk 2]
```diff
@@ -0,0 +1,6 @@
+abstract class RegistrationService {
+  Future<void> registerUser(
```
Comment (zddarova) [inline lib/data/services/registration_service.dart:2]: same with return type here

File: lib/data/services/registration_service_impl.dart [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import 'package:firebase_auth/firebase_auth.dart';
+import 'package:water_tracker/data/services/registration_service.dart';
+
+class RegistrationServiceFirebaseImpl extends RegistrationService {
+  final credential = FirebaseAuth.instance;
+
+  @override
+  Future<void> registerUser(String password, String email) async {
+    credential.createUserWithEmailAndPassword(
```
Comment (zddarova) [inline lib/data/services/registration_service_impl.dart:9]: you are not awaiting it

Comment (zddarova) [reply lib/data/services/registration_service_impl.dart:9]: also, createUserWithEmailAndPassword method returns some model. Check it out in order to find out is it contains any useful information for you. For example, it is contains token.

File: lib/data/services/registration_service_impl.dart [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import 'package:firebase_auth/firebase_auth.dart';
+import 'package:water_tracker/data/services/registration_service.dart';
+
+class RegistrationServiceFirebaseImpl extends RegistrationService {
+  final credential = FirebaseAuth.instance;
```
Comment (zddarova) [inline lib/data/services/registration_service_impl.dart:5]: not sure about credential name. Mby auth?

File: lib/get_it.dart [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import 'package:get_it/get_it.dart';
+import 'package:water_tracker/data/repository/repository.dart';
+import 'package:water_tracker/data/repository/repository_impl..dart';
+import 'package:water_tracker/data/services/registration_service.dart';
+import 'package:water_tracker/data/services/registration_service_impl.dart';
+
+final getIt = GetIt.instance;
+
+void setupServicesLocator() {
+  final registrationService = RegistrationServiceFirebaseImpl();
```
Comment (zddarova) [inline lib/get_it.dart:10]: could get rid of it as variable

File: lib/main.dart [hunk 6]
```diff
@@ -4,12 +4,14 @@ import 'package:water_tracker/water_app.dart';
 import 'generated/codegen_loader.g.dart';
 import 'package:firebase_core/firebase_core.dart';
 import 'firebase_options.dart';
+import 'get_it.dart';
 
 void main() async {
   WidgetsFlutterBinding.ensureInitialized();
   await Firebase.initializeApp(
-    options: DefaultFirebaseOptions.currentPlatform, // todo: what is this?
+    options: DefaultFirebaseOptions.currentPlatform,
```
Comment (zddarova) [inline lib/main.dart:12]: happy that you find out what is this

File: lib/presentation/screens/registration_screen/bloc/registration_state.dart [hunk 7]
```diff
@@ -0,0 +1,3 @@
+abstract class RegistrationState {}
+
+class InitialRegistrationState extends RegistrationState {}
```
Comment (zddarova) [inline lib/presentation/screens/registration_screen/bloc/registration_state.dart:3]: PR is called implement BL. And it is not implementing business logic...

File: lib/presentation/screens/registration_screen/bloc/registration_bloc.dart [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+import 'package:water_tracker/presentation/screens/registration_screen/bloc/registration_events.dart';
+import 'package:water_tracker/presentation/screens/registration_screen/bloc/registration_state.dart';
+
+class RegistrationBloc extends Bloc<RegistrationEvent, RegistrationState> {
+  Repository repository;
+
+  RegistrationBloc(this.repository) : super(InitialRegistrationState()) {
+    on<CreateUserEvent>((event, emit) async {
+      try {
+        repository.registerUser(event.password, event.email);
```
Comment (zddarova) [inline lib/presentation/screens/registration_screen/bloc/registration_bloc.dart:12]: PR is called implement BL. And it is not implementing business logic...

File: lib/presentation/screens/registration_screen/registration_screen.dart [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+import 'package:water_tracker/get_it.dart';
+import 'package:water_tracker/presentation/screens/registration_screen/bloc/registration_bloc.dart';
+import 'registration_layout.dart';
+
+class RegistrationScreen extends StatelessWidget {
+  const RegistrationScreen({Key? key}) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    return SafeArea(
```
Comment (zddarova) [inline lib/presentation/screens/registration_screen/registration_screen.dart:13]: stop wrapping every screen in safearea. Wrap all app 1 time.

File: lib/presentation/widgets/registration_form.dart [hunk 10]
```diff
@@ -57,21 +59,9 @@ class _MyLogFormWidgetState extends State<RegisterForm> {
           spacer,
           CustomButton(
             onPressed: () async {
-              try {
```
Comment (zddarova) [inline lib/presentation/widgets/registration_form.dart:60]: WHY THE FUCK IS IT HERE????
NEED TO DISCUSS

File: lib/presentation/widgets/registration_form.dart [hunk 11]
```diff
@@ -57,21 +59,9 @@ class _MyLogFormWidgetState extends State<RegisterForm> {
...
-              } on FirebaseAuthException catch (e) {
-                if (e.code == 'weak-password') {
-                  print('The password provided is too weak.');
-                } else if (e.code == 'email-already-in-use') {
-                  print('The account already exists for that email.');
-                }
-              } catch (e) {
-                print(e);
-              }
+              context.read<RegistrationBloc>().add(CreateUserEvent(_pass.text, _email.text));
```
Comment (zddarova) [inline lib/presentation/widgets/registration_form.dart:62]: there is always login first, password second

File: lib/presentation/widgets/registration_form.dart [hunk 12]
```diff
@@ -57,21 +59,9 @@ class _MyLogFormWidgetState extends State<RegisterForm> {
...
+              print('New creation User');
+
               // TODO: call something on validation
               if (formKey.currentState!.validate()) {}
```
Comment (zddarova) [inline lib/presentation/widgets/registration_form.dart:66]: mby need to register user only if data is valid?

**Review by zddarova — CHANGES_REQUESTED** [review]

**Comment by zddarova** [comment]
@k1mka besides all comments, you implelented here NOT business logic, like stated in branch and PR names, but data layer and started to implement BL. PR is not done yet, and it is need to be few smaller PRs.

**Comment by zddarova** [comment]
@k1mka and you are bad at Jira ticket management. Need to discuss https://evil-monkey.atlassian.net/browse/WT-5

File: lib/data/services/registration_service_impl.dart (hunk 3 above)
Comment (k1mka) [reply lib/data/services/registration_service_impl.dart:9]: Yes, it contains a token and user id

File: lib/data/services/registration_service_impl.dart (hunk 4 above)
Comment (k1mka) [reply lib/data/services/registration_service_impl.dart:5]: I like my option more

File: lib/presentation/screens/registration_screen/bloc/registration_bloc.dart (hunk 8 above)
Comment (k1mka) [reply lib/presentation/screens/registration_screen/bloc/registration_bloc.dart:12]: creating a block is part of the implementation of business logic

File: lib/presentation/screens/registration_screen/registration_screen.dart (hunk 9 above)
Comment (k1mka) [reply lib/presentation/screens/registration_screen/registration_screen.dart:13]: I tried but it caused a bunch of errors

File: lib/presentation/widgets/registration_form.dart (hunk 10 above)
Comment (k1mka) [reply lib/presentation/widgets/registration_form.dart:60]: I deleted async and try

File: lib/presentation/widgets/registration_form.dart (hunk 11 above)
Comment (k1mka) [reply lib/presentation/widgets/registration_form.dart:62]: ok

File: lib/presentation/widgets/registration_form.dart (hunk 12 above)
Comment (k1mka) [reply lib/presentation/widgets/registration_form.dart:66]: Yes, I fixed that yesterday

File: lib/presentation/screens/registration_screen/bloc/registration_bloc.dart [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+import 'package:water_tracker/presentation/screens/registration_screen/bloc/registration_state.dart';
+
+class RegistrationBloc extends Bloc<RegistrationEvent, RegistrationState> {
+  Repository repository;
+
+  RegistrationBloc(this.repository) : super(InitialRegistrationState()) {
+    on<CreateUserEvent>((event, emit) async {
+      try {
+        emit(LoadingRegistrationState());
+        await repository.registerUser(event.password, event.email);
```
Comment (zddarova) [inline lib/presentation/screens/registration_screen/bloc/registration_bloc.dart:13]: you could get here a bool and emit error state if user is not registered by any reason

File: lib/presentation/screens/registration_screen/registration_layout.dart [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+import 'package:water_tracker/presentation/widgets/already_have_an_account.widget.dart';
+import 'package:water_tracker/presentation/widgets/logo_widget.dart';
+import 'package:water_tracker/presentation/widgets/personal_image_widget.dart';
+import 'package:water_tracker/presentation/widgets/privacy_policy_and_terms_widget.dart';
+import 'package:water_tracker/presentation/widgets/registration_form.dart';
+
+class RegistrationLayout extends StatefulWidget {
+  const RegistrationLayout({Key? key}) : super(key: key);
+
+  static const spaces = SizedBox(height: 24);
```
Comment (zddarova) [inline lib/presentation/screens/registration_screen/registration_layout.dart:15]: could be moved to state class

File: lib/presentation/screens/registration_screen/registration_layout.dart [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+  static const paddingHorizontal = EdgeInsets.symmetric(horizontal: 24.0);
```
Comment (zddarova) [inline lib/presentation/screens/registration_screen/registration_layout.dart:16]: could be moved to state class

File: lib/presentation/screens/registration_screen/registration_layout.dart [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+class _RegistrationLayoutState extends State<RegistrationLayout> {
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      body: Padding(
+        padding: RegistrationLayout.paddingHorizontal,
+        child: BlocConsumer<RegistrationBloc, RegistrationState>(
+          listener: (context, state) {
+            if (state is ErrorRegistrationState) {
+              ScaffoldMessenger.of(context).showSnackBar(const SnackBar(content: Text('Error')));
```
Comment (zddarova) [inline lib/presentation/screens/registration_screen/registration_layout.dart:31]: move error to localization

File: lib/presentation/widgets/logo_form.dart [hunk 17]
```diff
@@ -51,6 +51,7 @@ class _MyLogFormWidgetState extends State<LogoForm> {
               if (formKey.currentState!.validate()) {}
             },
             text: LocaleKeys.sing_in.tr(),
+            isEnabled: true,
```
Comment (zddarova) [inline lib/presentation/widgets/logo_form.dart:54]: it has a default value of true. no need to pass it here

File: lib/presentation/widgets/registration_form.dart [hunk 18]
```diff
@@ -26,6 +31,13 @@ class _MyLogFormWidgetState extends State<RegisterForm> {
     super.initState();
   }
 
+  @override
+  void dispose() {
+    _pass.dispose();
```
Comment (zddarova) [inline lib/presentation/widgets/registration_form.dart:36]: 👍

File: lib/presentation/widgets/registration_form.dart [hunk 19]
```diff
@@ -55,14 +67,17 @@ class _MyLogFormWidgetState extends State<RegisterForm> {
           ),
           spacer,
           CustomButton(
-            onPressed: () async {
-              // TODO: call something on validation
-              if (formKey.currentState!.validate()) {}
+            isEnabled: widget.isButtonEnabled,
+            onPressed: () {
```
Comment (zddarova) [inline lib/presentation/widgets/registration_form.dart:71]: mby pass callbalck from outside?

**Review by zddarova — CHANGES_REQUESTED** [review]

File: lib/presentation/widgets/logo_form.dart (hunk 17 above)
Comment (k1mka) [reply lib/presentation/widgets/logo_form.dart:54]: I will change it when I start to work  with the login

**Review by zddarova — APPROVED** [review]

**Pull request merged**